[draft][a11y]make semanticsflags a struct/class instead of a bitmask

hannah-hyj opened the pull request:
* Embedder.h
  add a struct FlutterSemanticsFlags 
  add FlutterSemanticsFlags* to FlutterSemanticsNode2

 * engine/src/flutter/lib/ui/semantics.dart
  add a SemanticsFlags class for flags.

* web engine 
  same SemanticsFlags class for web engine.

 * JNI
    convert the flags struct to int64 bitmask and save it into buffer
    read it as long on java side.

 * SemanticsUpdateBuilder.updateNode
    pass a list of bools

 * Framework 
    use the SemanticsFlags class instead of int 

 * flutter_tester
    keep the old api to avoid breaking changes, will migrate them later.


## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

hannah-hyj committed:
2

Update SemanticsObjectTest.mm

2

Update semantics_node.h

make semanticsflags a struct

yjbanov commented on engine/src/flutter/shell/platform/common/accessibility_bridge.cc:
```diff
@@ -311,43 +311,41 @@ void AccessibilityBridge::ConvertFlutterUpdate(const SemanticsNode& node,
-  FlutterSemanticsFlag flags = node.flags;
-  if (flags & FlutterSemanticsFlag::kFlutterSemanticsFlagIsButton) {
+  FlutterSemanticsFlags flags = *(node.flags);
+  if (flags.isButton) {
```
Wow, this is so much cleaner than the previous code!

yjbanov commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:
```diff
@@ -460,7 +460,9 @@ - (void)testReplacedSemanticsDoesNotCleanupChildren {
     new_node.id = 1;
     new_node.rect = SkRect::MakeXYWH(0, 0, 100, 200);
-    new_node.flags = static_cast<int32_t>(flutter::SemanticsFlags::kHasImplicitScrolling);
+    FlutterSemanticsFlags flags;
```
Is this variable used for anything? (can't find any usages)

yjbanov commented on engine/src/flutter/shell/platform/darwin/ios/framework/Source/accessibility_bridge_test.mm:
```diff
@@ -460,7 +460,9 @@ - (void)testReplacedSemanticsDoesNotCleanupChildren {
     new_node.rect = SkRect::MakeXYWH(0, 0, 100, 200);
-    new_node.flags = static_cast<int32_t>(flutter::SemanticsFlags::kHasImplicitScrolling);
+    FlutterSemanticsFlags flags;
+    new_node.flags = flutter::SemanticsFlags2{};
```
Can we make it so that a new node is initialized with an empty set of flags as part of construction? This would remove the `new_node.flags = flutter::SemanticsFlags2{}` boilerplate that appears a lot here and below.

yjbanov commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/AccessibilityBridgeMacTest.mm:
```diff
@@ -99,8 +99,9 @@ @implementation AccessibilityBridgeTestViewController
+  FlutterSemanticsFlags flags;
   root.id = 0;
-  root.flags = static_cast<FlutterSemanticsFlag>(0);
+  root.flags2 = &flags;
```
nit: the `flags` variable seems superfluous if its only used to pass an empty set of flags to `root.flags2`.

hannah-hyj replied:
Yes, but i want to pass a pointer, so I'm defining the flags first. I don't know if there's a better way to do it.

yjbanov commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:
```diff
@@ -311,8 +311,10 @@ @implementation MockableFlutterEngine
   // Send flutter semantics updates.
   FlutterSemanticsNode2 root;
+  FlutterSemanticsFlags flags;
+  FlutterSemanticsFlags child_flags;
```
same nit here

chunhtai commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -1594,6 +1628,9 @@ typedef struct {
   const FlutterStringAttribute** decreased_value_attributes;
+  // The set of semantics flags associated with this node. Prefer to use this
+  // over `flags`.
+  FlutterSemanticsFlags* flags2;
```
I vaguely remember we have to be careful about using pointer where the value structure may expand in the future, but i can't remember exactly what may have gone wrong if we don't. cc @loic-sharma do you think this will be ok? We are going to add more bool to the struct FlutterSemanticsFlags in the future.

chunhtai commented on engine/src/flutter/shell/platform/embedder/embedder_semantics_update.cc:
```diff
@@ -26,6 +27,114 @@ EmbedderSemanticsUpdate::EmbedderSemanticsUpdate(
   };
 }
 
+FlutterSemanticsFlag SemanticsFlagtoInt(const SemanticsFlags2& flags) {
```
if this is only needed for jni, this method should probably be implement in jni instead of here.

hannah-hyj replied:
This is not for jni, 
in embedder.h I added a struct `flags2` but kept the old int `flags`.
this is used to set the value for the old int `flags`  in embedder.h. 

I actually don't know if setting the value for it is necessary 😅 , do you know the answer?

chunhtai replied:
we should make sure old semantics flag still reflects in the old int in embedder.h

hannah-hyj replied:
just for my curiosity, what will happen if we don't set the value for the  old semantics flag?

yjbanov commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -258,6 +258,40 @@ typedef enum {
+  bool hasSelectedState;
+  bool hasRequiredState;
+  bool isRequired;
+} FlutterSemanticsFlags;
```
What's an ergonomic/idiomatic way to guarantee that the struct is initialized with all fields set to `false`. AFAIK, C does not generally guarantee that the memory is zeroed out upon initialization. Simply creating a variable of type `FlutterSemanticsFlags` may end up with some random subset of flags set to `true`.

hannah-hyj replied:
looks like we can do FlutterSemanticsFlags{0} to initialize all fields to 0. updated tests.

yjbanov commented on engine/src/flutter/lib/ui/semantics.dart:
```diff
@@ -962,6 +961,233 @@ class SemanticsFlag {
+  bool isExpanded;
+  bool hasSelectedState;
+  bool hasRequiredState;
+  bool isRequired;
```
Can we make all fields `final`? Otherwise, `operator ==` and `hashCode` would misbehave, and sharing values of this class would be difficult because it's easy to mutate an instance that should not be mutated. If it's convenient to have a mutable variant of this class, consider using the builder-value pattern. The builder - `SemanticsFlagsBuilder` - is mutable, it's not expected to be shared, and it doesn't have `operator==` or `hashCode`. However, the value - `SemanticsFlags` - is immutable, safe to share, and it provides `operator==` and `hashCode`.

hannah-hyj committed:
resolve some comments

resolve some comments

lint

web

update some tests

Update accessibility_bridge_test.mm

resolve some comments

hannah-hyj committed:
Update semantics.dart

web

hannah-hyj committed:
fix tests

fix tests

Update accessibility_bridge.cc

hannah-hyj committed:
fix more tests

hannah-hyj committed:
fix more tests

hannah-hyj commented:
I think this PR is getting too big, i might split it into 2 or 3 parts.

hannah-hyj committed:
fix tests

hannah-hyj committed:
Update semantics_tester.dart

hannah-hyj commented:
I haven't fixed all the tests but I think this PR is getting too big and hard to review and land, I will split it into a few parts. Starting with the dart:ui part 🙂
